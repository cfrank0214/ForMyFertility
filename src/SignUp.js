
import React, { Component } from 'react';
import {
  TextInput,
  Button,
  StyleSheet,
  Text,
  View
} from 'react-native';

import { Auth } from 'aws-amplify' 


export default class SignUp extends Component {
  state = {
    username: '',
    password: '',
    phone_number: '',
    email: '',
    confirmationCode: ''
  }
  onChangeText(key,value) {
    this.setState({
      [key]: value
    })
  }
  signUp() {
    Auth.signUp({
      username: this.state.username,
      password: this.state.password,
      attributes: {
        email: this.state.email,
        phone_number: this.state.phone_number
      }
    })
    .then(() => console.log('successful sign up!'))
    .catch(err => console.log('error signing up!: ', err))
  }
  confirmSignUp() {
    Auth.confirmSignUp(this.state.username,this.state.confirmationCode)
    .then(() => console.log('successful confirming sign up!'))
    .catch(err => console.log('error confirming signing up!: ', err))
  }
  render() {
    return (
        <View style={styles.container}>
          <TextInput
          onChangeText={value => this.onChangeText('username', value)}
          style={styles.input}
          placeholder='username' 
          />
          <TextInput
          onChangeText={value => this.onChangeText('password', value)}
          style={styles.input}
          secureTextEntry={true}
          placeholder='password' 
          />
          <TextInput
          onChangeText={value => this.onChangeText('phone_number', value)}
          style={styles.input}
          placeholder='phone_number' 
          />
          <TextInput
          onChangeText={value => this.onChangeText('email', value)}
          style={styles.input}
          placeholder='email' 
          />
          <Button title='Sign Up' onPress={this.signUp.bind(this)} />
          <TextInput
          onChangeText={value => this.onChangeText('confirmationCode', value)}
          style={styles.input}
          placeholder='Confirmation Code' 
          />
          <Button title='Confirm Sign Up' onPress={this.confirmSignUp.bind(this)} />
        </View>
    );
  }
}



const styles = StyleSheet.create({
  input: {
    height: 50,
    borderBottomWidth: 2,
    borderBottomColor: '#2196F3',
    margin: 10
  },
  container: {
    flex: 1,
    justifyContent: 'center',
    backgroundColor: '#F5FCFF',
  },
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
  },
  instructions: {
    textAlign: 'center',
    color: '#333333',
    marginBottom: 5,
  },
});
