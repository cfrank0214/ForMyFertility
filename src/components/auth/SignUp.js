import React, { Component } from 'react';
import {
  Container,
  Body,
  Button,
  Form,
  Text,
  Header,
  Content,
  Item,
  Input,
  Label
} from 'native-base';
import PhoneInput from 'react-native-phone-input'
import { Col, Row, Grid } from 'react-native-easy-grid';
import { Auth, Logger } from 'aws-amplify';

const logger = new Logger('SignUp');

export default class SignUp extends Component {
  constructor(props) {
    super(props);
    this.signUp = this.signUp.bind(this);
    this.changeState = this.changeState.bind(this);
    this.state = { 
      error: null,
      phone_number: '',
    }
  }

  changeState(state, data) {
    const { onStateChange } = this.props;
    if (onStateChange) {
      onStateChange(state, data);
    }
  }

  onChangeText(key, value) {
    this.setState({
      [key]: value
    })
  }

  signUp() {
    const { username, password, email, phone_number } = this.state;
    console.log(`${JSON.stringify(this.state)}`)
    logger.info('sign up with ' + username);
    Auth.signUp({
      username: username,
      password: password,
      phone_number: phone_number,
      attributes: {
        email: email,
        phone_number: phone_number
      },
    })
      .then(() => this.signUpSuccess(username))
      .catch(err => this.signUpError(err));
  }
  signUpSuccess(username) {
    logger.info('sign up success with ' + username);
    this.setState({ error: null });

    this.changeState('confirmSignUp', username);
  }

  signUpError(err) {
    logger.info('sign up error', err);
    let message = err.message || err;
    if (message.startsWith('Invalid phone number')) {
      // reference: https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-attributes.html
      message = 'Phone numbers must follow these formatting rules: A phone number must start with a plus (+) sign, followed immediately by the country code. A phone number can only contain the + sign and digits. You must remove any other characters from a phone number, such as parentheses, spaces, or dashes (-) before submitting the value to the service. For example, a United States-based phone number must follow this format: +14325551212.'
    }
    this.setState({ error: message });
  }

  render() {
    const { authState } = this.props;
    if (authState !== 'signUp') { return null; }

    const { error } = this.state;

    return (
      <Content>
        <Grid>
          <Form>
            <Item stackedLabel style={{ marginTop: 30 }}>
              <Label>Email (Username)</Label>
              <Input
                type="email"
                autoCapitalize="none"
                onChangeText={(email) => this.setState({ username: email, email: email })}
                autoFocus
              />
            </Item>
            <Item stackedLabel last>
              <Label>Password</Label>
              <Input
                type="password"
                secureTextEntry
                onChangeText={(password) => this.setState({ password })}
              />
            </Item>
            <Item stackedLabel style={{ marginTop: 30 }}>
              <Label>Phone number</Label>
              <PhoneInput
                style={{ margin: 10 }}
                ref='phone'
                value={this.state.phone_number}
                onChangeText={(val) => {
                  if (this.state.phone_number === '') {
                    // render UK phone code by default when Modal is not open
                    this.onChangeText('phone_number', defaultCode + val)
                  } else {
                    // render country code based on users choice with Modal
                    this.onChangeText('phone_number', val)
                  }
                }}
              />
            </Item>
            <Button block success onPress={() => { this.signUp() }}>
              <Text>
                Create account
                  </Text>
            </Button>
            {error &&
              <Text>{error}</Text>
            }
            <Row>
              <Button transparent onPress={() => this.changeState('signIn')}>
                <Text>
                  Back to sign in
                  </Text>
              </Button>
              <Button transparent onPress={() => this.changeState('confirmSignUp')} >
                <Text>
                  Confirm a code
                  </Text>
              </Button>
            </Row>
          </Form>
        </Grid>
      </Content>
    )
  }
}