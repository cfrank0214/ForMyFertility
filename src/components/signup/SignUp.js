
import React, { Component } from 'react';
import {
  Container,
  Button,
  Text,
  Body,
  Header,
  Content,
  Form,
  Item,
  Input,
  Label
} from 'native-base';

import { Auth } from 'aws-amplify'

export default class SignUp extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isLoading: false,
      email: "",
      password: "",
      confirmPassword: "",
      confirmationCode: "",
      newUser: null
    };
  }

  validateForm() {
    return (
      this.state.email.length > 0 &&
      this.state.password.length > 0 &&
      this.state.password === this.state.confirmPassword
    );
  }

  validateConfirmationForm() {
    return this.state.confirmationCode.length > 0;
  }

  handleChange = event => {
    this.setState({
      [event.target.id]: event.target.value
    });
  }

  handleSubmit = async event => {
    event.preventDefault();

    this.setState({ isLoading: true });

    try {
      const newUser = await Auth.signUp({
        username: this.state.email,
        password: this.state.password
      });
      this.setState({
        newUser
      });
    } catch (e) {
      alert(e.message);
    }

    this.setState({ isLoading: false });
  }

  handleConfirmationSubmit = async event => {
    event.preventDefault();

    this.setState({ isLoading: true });

    try {
      await Auth.confirmSignUp(this.state.email, this.state.confirmationCode);
      await Auth.signIn(this.state.email, this.state.password);

      this.props.screenProps.authenticate(true)
      this.props.navigation.navigate('FertilityForm');

    } catch (e) {
      alert(e.message);
      this.setState({ isLoading: false });
    }
  }

  renderConfirmationForm() {
    return (
      <Container>
        <Header />
        <Content padder>
          <Form>
            <Body style={{ marginTop: 10 }}>
              <Text>
                We take securing and protecting
                your information very seriously.
                Which is why we're asking you to
                create a free account before we
                ask you private questions about your
                body and fertility history.
            </Text>
            </Body>
            <Item stackedLabel style={{ marginTop: 30 }}>
              <Label>Confirmation Code</Label>
              <Input
                onChange={this.handleChange}
                value={this.state.confirmationCode}
              />
            </Item>

          </Form>
          <Button
            block
            success
            disabled={!this.validateForm()}
            isLoading={this.state.isLoading}
            onPress={this.handleConfirmationSubmit}
          >
            <Text>
              Verify
            </Text>
          </Button>
        </Content>
      </Container>
    );
  }

  renderForm() {
    return (
      <Container>
        <Header />
        <Content padder>
          <Form>
            <Body style={{ marginTop: 10 }}>
              <Text>
                We take securing and protecting
                your information very seriously.
                Which is why we're asking you to
                create a free account before we
                ask you private questions about your
                body and fertility history.
            </Text>
            </Body>
            <Item stackedLabel style={{ marginTop: 30 }}>
              <Label>Email</Label>
              <Input
                keyboardType="email-address"
                autoCorrect={false}
                autoCapitalize="none"
                onChangeText={(email) => { this.setState({ email }); }}
                value={this.state.email}
              />
            </Item>
            <Item stackedLabel last>
              <Label>Password</Label>
              <Input
                secureTextEntry
                onChangeText={(password) => { this.setState({ password }); }}
                value={this.state.password}
              />
            </Item>
            <Item stackedLabel last>
              <Label>Confirm Password</Label>
              <Input
                secureTextEntry
                onChangeText={(confirmPassword) => { this.setState({ confirmPassword }); }}
                value={this.state.confirmPassword}
              />
            </Item>
          </Form>
          <Button
            block
            success
            disabled={!this.validateForm()}
            isLoading={this.state.isLoading}
            onPress={this.handleSubmit}
          >
            <Text>
              Signup
              </Text>
          </Button>
        </Content>
      </Container>
    );
  }

  render() {
    return (
      <Container className="Signup">
        {this.state.newUser === null
          ? this.renderForm()
          : this.renderConfirmationForm()}
      </Container>
    );
  }
}