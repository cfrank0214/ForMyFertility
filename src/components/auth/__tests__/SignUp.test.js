/**
 * @jest-environment jsdom
 */
import React from 'react'
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import SignUp from '../SignUp';
import {
  Container,
  Body,
  Button,
  Form,
  Text,
  Header,
  Content,
  Item,
  Input,
  Label
} from 'native-base';
import { Col, Row, Grid } from 'react-native-easy-grid';

describe('>>>SignUp --- Shallow Render REACT COMPONENTS', () => {
  let wrapper

  beforeEach(() => {

    const props = {
      inputs: jest.fn(),
      state: jest.fn(),
      authState: 'signUp'
    }

    wrapper = shallow(<SignUp {...props} />)

  })

  it('renders correctly', () => {
    const tree = renderer.create(
      <SignUp />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('+++ render the DUMB component', () => {
    expect(wrapper.length).toEqual(1)
  });
  it('The <Button /> component exists', () => {
    expect(wrapper.find(Button).exists()).toBe(true);
  });
  it('contains 3 <Button /> components', () => {
    expect(wrapper.find(Button).length).toBe(3);
  });
  it('contains 1 <Content /> component', () => {
    expect(wrapper.find(Content).length).toBe(1);
  });
  it('contains 1 <Grid /> component', () => {
    expect(wrapper.find(Grid).length).toBe(1);
  });
  it('contains 1 <Form /> component', () => {
    expect(wrapper.find(Form).length).toBe(1);
  });
  it('contains 3 <Item /> components', () => {
    expect(wrapper.find(Item).length).toBe(3);
  });
  it('contains 1 <Row /> component', () => {
    expect(wrapper.find(Row).length).toBe(1);
  });
});