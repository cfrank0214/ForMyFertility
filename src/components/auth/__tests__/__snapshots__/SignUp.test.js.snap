// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>SignUp --- Shallow Render REACT COMPONENTS renders correctly 1`] = `null`;
